src/bun.js/api/bun/udp_socket.zig

```diff
@@ -0,0 +1,751 @@
+    const callback = udpSocket.config.on_data;
+    if (callback == .zero) return;
+
+    const globalThis = udpSocket.globalThis;
+
+    var i: c_int = 0;
+    while (i < packets) : (i += 1) {
+        const peer = buf.getPeer(i);
+
+        var addr_buf: [INET6_ADDRSTRLEN + 1:0]u8 = undefined;
+        var hostname: ?[*:0]const u8 = null;
+        var port: u16 = 0;
+
+        switch (peer.family) {
+            std.os.AF.INET => {
+                const peer4: *std.os.sockaddr.in = @ptrCast(peer);
+                hostname = inet_ntop(peer.family, &peer4.addr, &addr_buf, addr_buf.len);
+                port = ntohs(peer4.port);
+            },
+            std.os.AF.INET6 => {
+                const peer6: *std.os.sockaddr.in6 = @ptrCast(peer);
+                hostname = inet_ntop(peer.family, &peer6.addr, &addr_buf, addr_buf.len);
+                port = ntohs(peer6.port);
+            },
+            else => continue,
+        }
+
+        if (hostname == null or port == 0) {
+            continue;
+        }
+
+        const slice = buf.getPayload(i);
+
+        const loop = udpSocket.vm.eventLoop();
+        loop.enter();
+        defer loop.exit();
+        udpSocket.js_refcount += 1;
+        defer udpSocket.js_refcount -= 1;
+
+        const result = callback.callWithThis(globalThis, udpSocket.thisValue, &[_]JSValue{
+            udpSocket.thisValue,
+            udpSocket.config.binary_type.toJS(slice, globalThis),
+            JSC.jsNumber(port),
+            JSC.ZigString.init(std.mem.span(hostname.?)).toValueAuto(globalThis),
+        });
+
+        if (result.toError()) |err| {
+            _ = udpSocket.callErrorHandler(.zero, &[_]JSValue{err});
+        }
+    }
+}
+
+pub const UDPSocketConfig = struct {
+    const This = @This();
+    const handlers = .{
+        .{ "data", "on_data" },
+        .{ "drain", "on_drain" },
+        .{ "error", "on_error" },
+    };
+
+    const ConnectConfig = struct {
+        port: u16,
+        address: [:0]u8,
+    };
+
+    hostname: [:0]u8,
+    connect: ?ConnectConfig = null,
+    port: u16,
+    binary_type: JSC.BinaryType = .Buffer,
+    on_data: JSValue = .zero,
+    on_drain: JSValue = .zero,
+    on_error: JSValue = .zero,
+
+    pub fn fromJS(globalThis: *JSGlobalObject, options: JSValue) ?This {
+        if (options.isEmptyOrUndefinedOrNull() or !options.isObject()) {
+            globalThis.throwInvalidArguments("Expected an object", .{});
+            return null;
+        }
+
+        const hostname = brk: {
+            if (options.getTruthy(globalThis, "hostname")) |value| {
+                if (!value.isString()) {
+                    globalThis.throwInvalidArguments("Expected \"hostname\" to be a string", .{});
+                    return null;
+                }
+                const str = value.toBunString(globalThis);
+                defer str.deref();
+                break :brk str.toOwnedSliceZ(default_allocator) catch bun.outOfMemory();
```

Comment:
These will leak on error

Instead, I suggest:
1) Move things which allocate to the end of this function
2) Assign them directly to `This` instead of a local variable 
3) `defer if (globalThis.hasPendingException()) config.deinit()`
4) `config.deinit` should call `unprotect` on the handlers to ensure we do not leak the callbacks